### PR TITLE
KTOR-572 Retry on HttpCode or network error

### PR DESCRIPTION
**Pull request by rsinukov**
API examples 
```
install(HttpRequestRetry) {
    retryOnServerErrors(3)
    exponentialDelay()
}
```
```
install(HttpRequestRetry) {
    retryOnException(3)
    constantDelay(millis = 5000)
}
```
```
install(HttpRequestRetry) {
    maxRetries = 3
    retryIf { request, response -> response.status.value.let { it in 500..599 } }
    retryOnExceptopmIf { request, cause -> cause is IoException }
}
```
Per request config
```
val request = HttpRequestBuilder {
    url(...)
    retry { noRetry() }
}
```

### CODE REVIEW

File: ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/HttpRequestRetry.kt [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+            val maxRetries = request.attributes.getOrNull(MaxRetriesPerRequestAttributeKey) ?: maxRetries
+            val delayMillis = request.attributes.getOrNull(RetryDelayPerRequestAttributeKey) ?: delayMillis
+
+            var call: HttpClientCall
+            while (true) {
+                val subRequest = prepareRequest(request)
+
+                val retryData = try {
+                    call = execute(subRequest)
+                    val retry = retryCount < maxRetries && shouldRetry(call.request, call.response)
```
Comment (e5l) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/HttpRequestRetry.kt:181]: I would extract the condition in a separate method

File: ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/HttpRequestRetry.kt [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+        public val response: HttpResponse?,
+        public val cause: Throwable?
+    )
+
+    private val shouldRetry: (HttpRequest, HttpResponse) -> Boolean = configuration.shouldRetry
+    private val shouldRetryOnException: (HttpRequestBuilder, Throwable) -> Boolean =
+        configuration.shouldRetryOnException
+    private val delayMillis: (Int) -> Long = configuration.delayMillis
+    private val delay: suspend (Long) -> Unit = configuration.delay
+    private var maxRetries: Int = configuration.maxRetries
```
Comment (e5l) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/HttpRequestRetry.kt:40]: `var` -> `val`

**Review by e5l — CHANGES_REQUESTED** [review]
LGTM, 
@hhariri could you check the public API docs?

File: ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/HttpRequestRetry.kt [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+
+            delayMillis {
+                millis + randomMs(randomizationMs)
+            }
+        }
+
+        /**
+         * Delay policy that implements [Exponential Backoff](https://en.wikipedia.org/wiki/Exponential_backoff)
+         * pattern. Retries will be calculated using `base ^ retryCount + [0..randomizationMs]`
+         */
```
Comment (andreyaksenov) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/HttpRequestRetry.kt:154]: > Specifies an exponential delay between retries, which is calculated using the [Exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff) algorithm.

File: ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/HttpRequestRetry.kt [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+            delayMillis { retry ->
+                val delay = minOf(base.pow(retry).toLong() * 1000L, maxDelayMs)
+                delay + randomMs(randomizationMs)
+            }
+        }
+
+        /**
+         * Function that awaits for specified amount of milliseconds. Uses [kotlinx.coroutines.delay] by default.
+         * Useful for tests.
+         */
```
Comment (andreyaksenov) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/HttpRequestRetry.kt:173]: > A function that waits for the specified amount of milliseconds. ...

File: ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/HttpRequestRetry.kt [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+        override fun install(plugin: HttpRequestRetry, scope: HttpClient) {
+            plugin.intercept(scope)
+        }
+    }
+}
+
+/**
+ * Configures [HttpRequestRetry] plugin on per request level
+ */
```
Comment (andreyaksenov) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/HttpRequestRetry.kt:268]: > Configures the [HttpRequestRetry] plugin on a per-request level.

**Review by andreyaksenov — CHANGES_REQUESTED** [review]

File: ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/HttpRequestRetry.kt [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+import io.ktor.events.*
+import io.ktor.util.*
+import kotlinx.coroutines.*
+import kotlin.coroutines.*
+import kotlin.math.*
+import kotlin.native.concurrent.*
+import kotlin.random.*
+
+/**
+ * Plugin that provides retry functionality. Default retry policy is no retry.
```
[the remark this answers is not in the document]
Comment (rsinukov) [reply ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/HttpRequestRetry.kt:20]: Updated. Yes, default means no configuration

**Review by e5l — APPROVED** [review]

**Review by andreyaksenov — APPROVED** [review]

**Pull request merged**